File: Packages/NMP/Sources/NMP/NIP25.swift
// Opaque typed NIP-25 target/draft projection (#155). Target qualification
// re-reads the canonical Rust store by event id; the caller's raw Row fields
// and sources are not trusted inputs. Draft bytes stay opaque for a closed
// publisher such as the upcoming NIP-29 publication gate.

import NMPFFI

/// One validated semantic NIP-25 reaction value.
public enum ReactionValue: Sendable, Hashable {
    case like
    case dislike
    case emoji(String)
    case customEmoji(shortcode: String, imageURL: String)

    func toFfi() -> FfiReactionValue {
        switch self {
        case .like:
            return .like
        case .dislike:
            return .dislike
        case .emoji(let value):
            return .emoji(value: value)
        case .customEmoji(let shortcode, let imageURL):
            return .customEmoji(shortcode: shortcode, imageUrl: imageURL)
        }
    }
}

/// Opaque capability proving that NMP qualified one canonical signed event
/// as a complete native-event reaction target.
public final class ReactionTarget: @unchecked Sendable {
    let ffi: FfiReactionTarget

    init(_ ffi: FfiReactionTarget) {
        self.ffi = ffi
    }
}

/// Opaque immutable unsigned protocol draft. It exposes no event kind, tags,
/// author, time, routing, signing, receipt, retry, or publication operation.
public final class ProtocolDraft: @unchecked Sendable {
    let ffi: FfiProtocolDraft

    init(_ ffi: FfiProtocolDraft) {
        self.ffi = ffi
    }
}

/// Typed failures from NIP-25 target qualification and draft composition.
public enum ReactionError: Error, Sendable, Equatable {
    case invalidEventID(got: String)
    case targetNotFound(eventID: String)
    case targetNotVerified(eventID: String)
    case canonicalLookupUnavailable(reason: String)
    case engineClosed
    case noActiveReactionAuthor
    case emptyEmoji
    case standardValueRequiresTypedVariant(got: String)
    case customEmojiRequiresMetadata(got: String)
    case invalidEmojiToken(got: String)
    case invalidCustomEmojiShortcode(got: String)
    case invalidCustomEmojiURL(got: String)

    init(_ ffi: FfiReactionError) {
        switch ffi {
        case .InvalidEventId(let got):
            self = .invalidEventID(got: got)
        case .TargetNotFound(let eventId):
            self = .targetNotFound(eventID: eventId)
        case .TargetNotVerified(let eventId):
            self = .targetNotVerified(eventID: eventId)
        case .CanonicalLookupUnavailable(let reason):
            self = .canonicalLookupUnavailable(reason: reason)
        case .EngineClosed:
            self = .engineClosed
        case .NoActiveReactionAuthor:
            self = .noActiveReactionAuthor
        case .EmptyEmoji:
            self = .emptyEmoji
        case .StandardValueRequiresTypedVariant(let got):
            self = .standardValueRequiresTypedVariant(got: got)
        case .CustomEmojiRequiresMetadata(let got):
            self = .customEmojiRequiresMetadata(got: got)
        case .InvalidEmojiToken(let got):
            self = .invalidEmojiToken(got: got)
        case .InvalidCustomEmojiShortcode(let got):
            self = .invalidCustomEmojiShortcode(got: got)
        case .InvalidCustomEmojiUrl(let got):
            self = .invalidCustomEmojiURL(got: got)
        }
    }
}

extension NMPEngine {
    /// Qualify `row.id` through NMP's canonical cache. Every other field on
    /// this caller-constructible row, including `sources`, is ignored.
    public func reactionTarget(for row: Row) throws -> ReactionTarget {
        do {
            return try ReactionTarget(ffi.reactionTarget(eventId: row.id))
        } catch let error as FfiReactionError {
            throw ReactionError(error)
        }
    }

    /// Compose one Rust-authored unsigned NIP-25 draft using NMP's active
    /// account and Rust-owned time. This does not publish the event.
    public func reactionDraft(
        target: ReactionTarget,
        value: ReactionValue
    ) throws -> ProtocolDraft {
        do {
            return try ProtocolDraft(
                ffi.reactionDraft(target: target.ffi, value: value.toFfi())
            )
        } catch let error as FfiReactionError {
            throw ReactionError(error)
        }
    }
}

File: Packages/NMPKotlin/src/main/kotlin/com/nmp/sdk/NIP25.kt
// Opaque typed NIP-25 target/draft projection (#155). Target qualification
// re-reads the canonical Rust store by event id; caller-provided Row fields
// and sources are never trusted inputs. Mirrors NIP25.swift.

package com.nmp.sdk

import uniffi.nmp_ffi.FfiProtocolDraft
import uniffi.nmp_ffi.FfiReactionException
import uniffi.nmp_ffi.FfiReactionTarget
import uniffi.nmp_ffi.FfiReactionValue

/** One validated semantic NIP-25 reaction value. */
sealed class ReactionValue {
    data object Like : ReactionValue()

    data object Dislike : ReactionValue()

    data class Emoji(val value: String) : ReactionValue()

    data class CustomEmoji(val shortcode: String, val imageUrl: String) : ReactionValue()

    internal fun toFfi(): FfiReactionValue =
        when (this) {
            Like -> FfiReactionValue.Like
            Dislike -> FfiReactionValue.Dislike
            is Emoji -> FfiReactionValue.Emoji(value)
            is CustomEmoji -> FfiReactionValue.CustomEmoji(shortcode, imageUrl)
        }
}

/** Opaque capability proving that NMP qualified one canonical signed event
 * as a complete native-event reaction target. */
class ReactionTarget internal constructor(
    internal val ffi: FfiReactionTarget,
)

/** Opaque immutable unsigned protocol draft. No event kind, tags, author,
 * time, routing, signing, receipt, retry, or publication operation is
 * exposed. */
class ProtocolDraft internal constructor(
    internal val ffi: FfiProtocolDraft,
)

/** Typed failures from NIP-25 target qualification and draft composition. */
sealed class ReactionError(message: String) : Exception(message) {
    data class InvalidEventId(val got: String) : ReactionError("invalid Nostr event id: $got")

    data class TargetNotFound(val eventId: String) :
        ReactionError("event $eventId is not in the canonical NMP store")

    data class TargetNotVerified(val eventId: String) :
        ReactionError("canonical row $eventId is not a verified signed event")

    data class CanonicalLookupUnavailable(val reason: String) :
        ReactionError("canonical target lookup unavailable: $reason")

    data object EngineClosed : ReactionError("engine already shut down")

    data object NoActiveAccount : ReactionError("NIP-25 draft requires an active account")

    data object EmptyEmoji : ReactionError("Unicode reaction must not be empty")

    data class StandardValueRequiresTypedVariant(val got: String) :
        ReactionError("$got must use the typed like/dislike variant")

    data class CustomEmojiRequiresMetadata(val got: String) :
        ReactionError("$got requires matching typed NIP-30 metadata")

    data class InvalidEmojiToken(val got: String) :
        ReactionError("$got contains whitespace or control characters")

    data class InvalidCustomEmojiShortcode(val got: String) :
        ReactionError("invalid custom emoji shortcode: $got")

    data class InvalidCustomEmojiUrl(val got: String) :
        ReactionError("custom emoji image URL is not HTTP(S): $got")

    companion object {
        internal fun from(ffi: FfiReactionException): ReactionError =
            when (ffi) {
                is FfiReactionException.InvalidEventId -> InvalidEventId(ffi.got)
                is FfiReactionException.TargetNotFound -> TargetNotFound(ffi.eventId)
                is FfiReactionException.TargetNotVerified -> TargetNotVerified(ffi.eventId)
                is FfiReactionException.CanonicalLookupUnavailable ->
                    CanonicalLookupUnavailable(ffi.reason)
                is FfiReactionException.EngineClosed -> EngineClosed
                is FfiReactionException.NoActiveAccount -> NoActiveAccount
                is FfiReactionException.EmptyEmoji -> EmptyEmoji
                is FfiReactionException.StandardValueRequiresTypedVariant ->
                    StandardValueRequiresTypedVariant(ffi.got)
                is FfiReactionException.CustomEmojiRequiresMetadata ->
                    CustomEmojiRequiresMetadata(ffi.got)
                is FfiReactionException.InvalidEmojiToken -> InvalidEmojiToken(ffi.got)
                is FfiReactionException.InvalidCustomEmojiShortcode ->
                    InvalidCustomEmojiShortcode(ffi.got)
                is FfiReactionException.InvalidCustomEmojiUrl -> InvalidCustomEmojiUrl(ffi.got)
            }
    }
}

/** Qualify [row.id] through NMP's canonical cache. Every other field on this
 * caller-constructible row, including [Row.sources], is ignored. */
@Throws(ReactionError::class)
fun NMPEngine.reactionTarget(row: Row): ReactionTarget =
    try {
        ReactionTarget(ffi.reactionTarget(row.id))
    } catch (error: FfiReactionException) {
        throw ReactionError.from(error)
    }

/** Compose one Rust-authored unsigned NIP-25 draft using NMP's active
 * account and Rust-owned time. This does not publish the event. */
@Throws(ReactionError::class)
fun NMPEngine.reactionDraft(
    target: ReactionTarget,
    value: ReactionValue,
): ProtocolDraft =
    try {
        ProtocolDraft(ffi.reactionDraft(target.ffi, value.toFfi()))
    } catch (error: FfiReactionException) {
        throw ReactionError.from(error)
    }
